hathor/event/websocket/factory.py

```diff
@@ -0,0 +1,99 @@
+from typing import Dict, Set
+from uuid import uuid4
+
+from autobahn.twisted.websocket import WebSocketServerFactory
+from structlog import get_logger
+from twisted.internet.task import LoopingCall
+
+from hathor.conf import HathorSettings
+from hathor.event.websocket.event_streaming import EventStreaming
+from hathor.event.websocket.protocol import HathorEventWebsocketProtocol
+from hathor.util import json_dumpb, json_loadb, reactor
+
+settings = HathorSettings()
+logger = get_logger()
+
+
+class EventWebsocketFactory(WebSocketServerFactory):
+    """ Websocket that will handle events
+    """
+
+    protocol = HathorEventWebsocketProtocol
+
+    def buildProtocol(self, id):
+        return self.protocol(self, id)
+
+    def __init__(self, event_storage):
+        self.connections: Set[HathorEventWebsocketProtocol] = set()
+        self.connections_to_stream_events: Dict[str, EventStreaming] = {}
+        super().__init__()
+
+        self.log = logger.new()
+        self.is_running = False
+        self.event_storage = event_storage
+
+        # A timer to periodically send new events to clients that sent start_streaming_events message
+        self._lc_send_events = LoopingCall(self._send_events_to_subscribed_clients)
+        self._lc_send_events.clock = reactor
+
+    def start(self):
+        self.is_running = True
+
+        # Start event sender
+        self._lc_send_events.start(settings.WS_SEND_EVENTS_INTERVAL, now=False)
+
+    def stop(self):
+        self.is_running = False
+
+        if self._lc_send_events.running:
+            self._lc_send_events.stop()
+
+    def on_client_close(self, connection: HathorEventWebsocketProtocol) -> None:
+        """Called when a ws connection is closed."""
+        # A connection closed before finishing handshake will not be in `self.connections`.
+        self.connections.discard(connection)
+        del self.connections_to_stream_events[connection.id]
+
+    def on_client_open(self, connection: HathorEventWebsocketProtocol) -> None:
+        """Called when a ws connection is opened (after handshaking)."""
+        connection.id = str(uuid4())
+        self.connections.add(connection)
+
+    def handle_message(self, connection: HathorEventWebsocketProtocol, data: bytes) -> None:
+        message = json_loadb(data)
```

Comment:
Should we map `message` to specific arguments for the handlers? I feel it's hard to understand the keys and their values for each type of message.

Reply:
This has been changed a lot. I think it has to be re-reviewed.